Clean branch with Lund Reweighting Code

easmith5 opened the pull request:
This branch matches gen-reco particles and cluster them into constituents of specific dark hadrons. These can then be reweighting using the Lund Reweighting code.

Commit:
Code to produce inputs needed for the lund reweighting procedure

kpedro88 commented on TreeMaker/interface/TreeMaker.h:
```diff
@@ -219,9 +219,10 @@ typedef TreeObject<vector<math::PtEtaPhiELorentzVector>,vector<math::PtEtaPhiELo
 typedef TreeObject<vector<vector<math::XYZVector>>,vector<vector<math::XYZVectorF>>> TreeObjectVVXYZVToF;
 typedef TreeObject<vector<vector<math::XYZPoint>>,vector<vector<math::XYZPointF>>> TreeObjectVVXYZPToF;
-//typedefs for doubles
+//typedefs for doublesx
```
revert typo?

kpedro88 commented on TreeMaker/interface/TreeMaker.h:
```diff
@@ -495,7 +512,7 @@ class TreeNestedVector : public TreeObject<std::vector<std::vector<BaseIn>>,std:
 		void SetConsumes(edm::ConsumesCollector && iC) override{
 			tok = iC.consumes<TopIn>(this->tag);
-		}
+ 		}
```
revert random space?

kpedro88 commented on TreeMaker/interface/TreeMaker.h:
```diff
@@ -594,3 +611,123 @@ template <>
+			    if (!associated && storeOffsets) { offsets.insert(std::end(offsets),subLength); }
+			    for(auto& isub : sub) {
+			        accum.insert(std::end(accum), std::begin(isub), std::end(isub));
+				if (!associated) offsets.insert(std::end(offsets), storeOffsets ? accum.size() : isub.size());
```
fix inconsistent usage of tabs vs. spaces for indentation (everywhere)

kpedro88 replied:
(and please follow existing usage in each file, i.e. don't change indentation in other lines not already modified)

kpedro88 commented on TreeMaker/interface/TreeMaker.h:
```diff
@@ -594,3 +611,123 @@ template <>
+		// From: https://stackoverflow.com/questions/17294629/merging-flattening-sub-vectors-into-a-single-vector-c-converting-2d-to-1d
+		void flatten(TripIn const& all, SubOut &accum, vector<int> &offsets) {
+			// Don't store any offsets if there are no sub-vectors
+			if (all.size() == 0) return;
```
`if (all.empty())` (more efficient)

kpedro88 replied:
this one is not fixed yet

easmith5 replied:
Ah looks like I fixed the one in the previously existing class only. Should be done now.

kpedro88 commented on TreeMaker/interface/TreeMaker.h:
```diff
@@ -594,3 +611,123 @@ template <>
+			}
+			// This protects against the case where there were >=1 empty sub-vectors, and only empty vectors
+			// Thus, nothing will be in the output (accum) vector, but the offsets would be all '0'
+			if (storeOffsets && !associated && accum.size() == 0) offsets.clear();
```
`accum.empty()`

kpedro88 commented on TreeMaker/interface/TreeMaker.h:
```diff
@@ -594,3 +611,123 @@ template <>
+				if (!associated) offsets.insert(std::end(offsets), storeOffsets ? accum.size() : isub.size());
+			    }
+			}
+			// This protects against the case where there were >=1 empty sub-vectors, and only empty vectors
```
to clarify this comment a little: it's the case where *everything* is empty *and* there are >=1 sub-vectors, right?

easmith5 replied:
This is actually just copied from the one-level nested class, as is most of this code, but I believe that is correct

kpedro88 commented on TreeMaker/interface/TreeMaker.h:
```diff
@@ -594,3 +611,123 @@ template <>
+		const string GetTripType() {
+			return "vector<" + GetTopType() + ">";
+		}
+                const string GetTopType() {
```
more spacing/indentation issues

kpedro88 commented on TreeMaker/python/makeTreeFromMiniAOD_cff.py:
```diff
@@ -1180,7 +1184,19 @@ def makeTreeFromMiniAOD(self,process):
             DarkGluonIDs = cms.vuint32(4900021),
             SMQuarkIDs = cms.vuint32(1,2,3,4,5,6),
+            nsubjettiness = cms.PSet(
+                useSoftDrop = cms.bool(True),
```
this parameter is not actually used for nsubjettiness, can be removed

kpedro88 commented on TreeMaker/python/makeTreeFromMiniAOD_cff.py:
```diff
@@ -1180,7 +1184,19 @@ def makeTreeFromMiniAOD(self,process):
+                nPass = cms.int32(999),
+                akAxesR0 = cms.double(999.0),
+            ),
+            #jetCollInstanceName=cms.string("SubJets"),                     
```
delete commented-out code

kpedro88 commented on TreeMaker/python/makeTreeFromMiniAOD_cff.py:
```diff
@@ -1197,9 +1213,41 @@ def makeTreeFromMiniAOD(self,process):
+                'HiddenSector:JetsDarkHadronJetsConstituents(JetsAK8_darkHadronJets_constituents)',
+                'HiddenSector:JetsDarkHadronJetsConstituentsGenMatchOnly(JetsAK8_darkHadronJets_constituentsGenMatchOnly)',
+                'HiddenSector:JetsDarkHadronJetsConstituentsNextDH(JetsAK8_darkHadronJets_constituentsNextDH)',
+
```
delete extra line (also below)

kpedro88 commented on TreeMaker/src/TreeMaker.cc:
```diff
@@ -129,6 +129,7 @@ TreeMaker::TreeMaker(const edm::ParameterSet& iConfig) :
 				case TreeTypes::t_vvdouble  : tmp = saveFloat ? (TreeObjectBase*)(new TreeNVDoubleToF(VarName,VarTitle,nestedVectors,storeOffsets,false,splitLevel)) : (TreeObjectBase*)(new TreeNVDouble(VarName,VarTitle,nestedVectors,storeOffsets,false,splitLevel)); break;
 				case TreeTypes::t_vvstring  : tmp = new TreeNVString(VarName,VarTitle,nestedVectors,storeOffsets,false,splitLevel); break;
 				case TreeTypes::t_vvlorentz : tmp = saveFloat ? (TreeObjectBase*)(new TreeNVLVToF(VarName,VarTitle,nestedVectors,storeOffsets,false,splitLevel)) : (TreeObjectBase*)(new TreeNVLV(VarName,VarTitle,nestedVectors,storeOffsets,false,splitLevel)); break;
+			        case TreeTypes::t_vvvlorentz: tmp = saveFloat ? (TreeObjectBase*)(new TreeNNVLVToF(VarName,VarTitle,nestedVectors,storeOffsets,false,splitLevel)) : (TreeObjectBase*)(new TreeNNVLV(VarName,VarTitle,nestedVectors,storeOffsets,false,splitLevel)); break;
```
tabs vs spaces

kpedro88 commented on Utils/src/HiddenSectorProducer.cc:
```diff
@@ -40,37 +44,50 @@ CandPtr daughter_noexcept(const T& mother, unsigned i) {
+    edm::ParameterSet njhConfig; 
     PidSet DarkSMediatorIDs_, DarkTMediatorIDs_, DarkQuarkIDs_, DarkHadronIDs_, DarkGluonIDs_, DarkStableIDs_, DarkFirstIDs_, SMQuarkIDs_;
+    std::unique_ptr<NjettinessHelper> beginStream(edm::StreamID) const {
+      return std::unique_ptr<NjettinessHelper>(new NjettinessHelper(njhConfig));
```
prefer `return std::make_unique<NjettinessHelper>(njhConfig);` (more concise, avoids `new ` which always looks like a memory leak even though it's not in this case)

kpedro88 commented on Utils/src/HiddenSectorProducer.cc:
```diff
@@ -40,37 +44,50 @@ CandPtr daughter_noexcept(const T& mother, unsigned i) {
 
-void HiddenSectorProducer::fillSet(PidSet& IDset, const std::string& name, const edm::ParameterSet& iConfig){
+void HiddenSectorProducer::fillSet(PidSet& IDset, const std::string& name, const edm::ParameterSet& iConfig)
+{
```
unnecessary change

kpedro88 commented on Utils/src/HiddenSectorProducer.cc:
```diff
@@ -196,19 +237,166 @@ double HiddenSectorProducer::calculateMT2(const edm::Handle<edm::View<reco::GenM
+      if(cands_out[j].size()==daughterPtrs.size()) continue;
+      bool keep = false; 
+      for(const auto& candJetDau : jet.daughterPtrVector()){ 
+	keep = candJetDau.key() == candPtr.key();
```
tabs vs spaces

kpedro88 commented on Utils/src/HiddenSectorProducer.cc:
```diff
@@ -196,19 +237,166 @@ double HiddenSectorProducer::calculateMT2(const edm::Handle<edm::View<reco::GenM
+
+    for(unsigned j = 0; j < h_jets->size(); ++j){
+      if (cands_out.size() <= j) cands_out.emplace_back();
+      if (pdgids_out.size() <= j) pdgids_out.emplace_back();
```
could one just do `resize(j)` for these two `out` vectors at the beginning of the function (outside of the loop) rather than emplacing one at a time?

kpedro88 commented on Utils/src/HiddenSectorProducer.cc:
```diff
@@ -196,19 +237,166 @@ double HiddenSectorProducer::calculateMT2(const edm::Handle<edm::View<reco::GenM
+	toMatch[pdgids.size()].genParts.push_back(genSubjetConstituents[i][j]);
+	toMatch[pdgids.size()].genIndexTop.push_back(i);
+	toMatch[pdgids.size()].genIndex.push_back(j);
+      }
```
further reduce duplication:
```cpp
int index = -1;
if (it != pdgids.end()) index = it - pdgids.begin();
else index = pdgids.size();
toMatch[index]...
```

kpedro88 commented on Utils/src/HiddenSectorProducer.cc:
```diff
@@ -196,19 +237,166 @@ double HiddenSectorProducer::calculateMT2(const edm::Handle<edm::View<reco::GenM
+    else if(std::find(pdgids.begin(), pdgids.end(), jetCandsPdgid[r]) == pdgids.end() ) {
+      toMatch[pdgids.size()].recoParts.push_back(jetCands[r]);
+      toMatch[pdgids.size()].recoIndex.push_back(r);
+    }
```
similarly here

kpedro88 commented on Utils/src/HiddenSectorProducer.cc:
```diff
@@ -196,19 +237,166 @@ double HiddenSectorProducer::calculateMT2(const edm::Handle<edm::View<reco::GenM
+    //Save matched indices to final array
+    for(unsigned i = 0; i < m.matchedIndex.size(); i++){
+      if (m.matchedIndex[i] != -1) {
+	indices[m.genIndexTop[i]][m.genIndex[i]] = m.recoIndex[m.matchedIndex[i]];
```
tabs vs spaces

kpedro88 commented on Utils/src/HiddenSectorProducer.cc:
```diff
@@ -196,19 +237,166 @@ double HiddenSectorProducer::calculateMT2(const edm::Handle<edm::View<reco::GenM
+      matchStageReco[leftovers.recoIndex[leftovers.matchedIndex[m]]] = 2;
+    }
+  }
+  recoMatched = tmpRecoMatched;
```
are these assignments actually needed, since we just return `indices` directly after this?

kpedro88 commented on Utils/src/HiddenSectorProducer.cc:
```diff
@@ -40,37 +44,50 @@ CandPtr daughter_noexcept(const T& mother, unsigned i) {
+    edm::EDGetTokenT<edm::View<reco::Candidate>> CandTok_;
+    edm::EDGetTokenT<std::vector<int>> GenIndexTok_;
     double coneSize_;
+    edm::ParameterSet njhConfig; 
```
member variables should end in _

kpedro88 commented on Utils/src/HiddenSectorProducer.cc:
```diff
@@ -367,13 +626,237 @@ void HiddenSectorProducer::produce(edm::StreamID, edm::Event& iEvent, const edm:
+      std::vector<CLorentzVector> tmp_darkHadronJets;
+      std::vector<std::vector<CLorentzVector> > tmp_darkHadronJets_constituents;
+      std::vector<std::vector<int> > tmp_darkHadronJets_ConstituentPdgid;
+      //std::vector<std::vector<CLorentzVector> > tmp_darkHadronSubjets_constituents;
```
delete commented-out code

kpedro88 commented on Utils/src/HiddenSectorProducer.cc:
```diff
@@ -367,13 +626,237 @@ void HiddenSectorProducer::produce(edm::StreamID, edm::Event& iEvent, const edm:
+        if (darkHadron!=nullptr){
+          darkHadronMap[darkHadron].push_back(dau);
+	  nDaus++;
+	}
```
tabs vs spaces

kpedro88 commented on Utils/src/HiddenSectorProducer.cc:
```diff
@@ -367,13 +626,237 @@ void HiddenSectorProducer::produce(edm::StreamID, edm::Event& iEvent, const edm:
+      for(const auto& entry : darkHadronMap){
+        tmp_darkHadrons.emplace_back(entry.first->pt(),entry.first->eta(),entry.first->phi(),entry.first->energy());
+        LorentzVector tmpjet;
+	std::vector<CLorentzVector> tmpjetconstituents;
```
tabs vs spaces

kpedro88 commented on Utils/src/HiddenSectorProducer.cc:
```diff
@@ -367,13 +626,237 @@ void HiddenSectorProducer::produce(edm::StreamID, edm::Event& iEvent, const edm:
+	  tmpjetconstituentspdgid.emplace_back(dau->pdgId());
+        }
+        tmp_darkHadronJets.emplace_back(tmpjet.pt(),tmpjet.eta(),tmpjet.phi(),tmpjet.energy());
+	tmp_darkHadronJets_constituents.push_back(tmpjetconstituents);
```
tabs vs spaces

kpedro88 commented on Utils/src/HiddenSectorProducer.cc:
```diff
@@ -367,13 +626,237 @@ void HiddenSectorProducer::produce(edm::StreamID, edm::Event& iEvent, const edm:
+    int genJetIndex = genIndex->at(recoJetIndex);
+    
+    Jets_nConstituents->push_back(jets_cands[recoJetIndex].size());
+    //Jets_constituents_pdgid->push_back(cands_pdgids[recoJetIndex]);
```
delete commented-out code

kpedro88 commented on Utils/src/HiddenSectorProducer.cc:
```diff
@@ -367,13 +626,237 @@ void HiddenSectorProducer::produce(edm::StreamID, edm::Event& iEvent, const edm:
+      std::vector<int> tmpMatchStage;
+      std::vector<int> tmpRecoPdgid;
+      for(unsigned j = 0; j < matchedRecoIndex[i].size(); j++){
+	if (matchedRecoIndex[i][j] != -1) {
```
tabs vs spaces

kpedro88 commented on Utils/src/HiddenSectorProducer.cc:
```diff
@@ -367,13 +626,237 @@ void HiddenSectorProducer::produce(edm::StreamID, edm::Event& iEvent, const edm:
+    int nRecoUnmatched = 0;
+    for(unsigned i = 0; i < recoMatched.size(); i++){
+      isAssigned.push_back(0);
+      //std::cout << recoMatched[i] << std::endl;
```
delete commented-out code

kpedro88 commented on Utils/src/HiddenSectorProducer.cc:
```diff
@@ -367,13 +626,237 @@ void HiddenSectorProducer::produce(edm::StreamID, edm::Event& iEvent, const edm:
+    std::vector<int> isAssigned;
+    int nRecoUnmatched = 0;
+    for(unsigned i = 0; i < recoMatched.size(); i++){
+      isAssigned.push_back(0);
```
could just do `std::vector<int> isAssigned(recoMatched.size(), 0)`? (before the loop)

kpedro88 commented on Utils/src/HiddenSectorProducer.cc:
```diff
@@ -367,13 +626,237 @@ void HiddenSectorProducer::produce(edm::StreamID, edm::Event& iEvent, const edm:
+	nRecoUnmatched++;
+
+	CLorentzVector recoPart = jets_cands[recoJetIndex][i];
+	std::vector<CLorentzVector> darkHadronJets = GenJets_darkHadronJets->at(genJetIndex);
```
`const std::vector<CLorentzVector>&` (avoid copy)

kpedro88 commented on Utils/src/HiddenSectorProducer.cc:
```diff
@@ -367,13 +626,237 @@ void HiddenSectorProducer::produce(edm::StreamID, edm::Event& iEvent, const edm:
+	}
+      }
+    }
+    double unmatchedFraction = nRecoUnmatched / jets_cands[recoJetIndex].size();
```
never actually used?

kpedro88 commented:
Also, the integration test failed.

Input file: https://cernbox.cern.ch/remote.php/dav/public-files/3xHyo11H3fAgCHZ/store.mc.RunIISummer20UL16MiniAODv2.QCD_HT2000toInf_TuneCP5_PSWeights_13TeV-madgraphMLM-pythia8.MINIAODSIM.106X_mcRun2_asymptotic_v17-v1.120000.0605C02A-A789-C548-8D73-9B7D0A4D5561_100evt.root

Test command:
```
cmsRun \
  $CMSSW_BASE/src/TreeMaker/Production/test/runMakeTreeFromMiniAOD_cfg.py \
  scenario=Summer20UL16 \
  dataset=file:store.mc.RunIISummer20UL16MiniAODv2.QCD_HT2000toInf_TuneCP5_PSWeights_13TeV-madgraphMLM-pythia8.MINIAODSIM.106X_mcRun2_asymptotic_v17-v1.120000.0605C02A-A789-C548-8D73-9B7D0A4D5561_100evt.root \
  outfile=Summer20UL16.QCD_HT2000toInf_TuneCP5_PSWeights_13TeV-madgraphMLM-pythia8 \
  numevents=100
```

Error message:
```
----- Begin Fatal Exception 08-May-2025 19:26:50 UTC-----------------------
An exception of category 'StdException' occurred while
   [0] Processing  Event run: 1 lumi: 37831 event: 44488172 stream: 0
   [1] Running path 'WriteTree'
   [2] Prefetching for module TreeMaker/'TreeMaker2'
   [3] Calling method for module HiddenSectorProducer/'HiddenSector'
Exception Message:
A std::exception was thrown.
vector::_M_range_check: __n (which is 1) >= this->size() (which is 0)
----- End Fatal Exception -------------------------------------------------
```

How to debug:
```
cd $CMSSW_BASE/src/TreeMaker/Utils
touch src/HiddenSectorProducer.cc
scram b USER_CXXFLAGS="-g"
cd $CMSSW_BASE/src/TreeMaker/Production/test/
cmsTraceExceptions cmsRun ...
```

Commit:
Fix a few things, and try to improve indentation and spacing

Commit:
One more fix

Commit:
apparently all of the TreeMaker.h fixes got lost, adding them back

Commit:
Fix comments I somehow didn't see in the first round

Commit:
All the other comments hidden in the PR

easmith5 commented:
Ok I think I've addressed these comments! I also fixed the failed integration test by wrapping the Lund related code in the signal_ requirement. I tried to fix the spacing - the HiddenSectorProducer class seems to not be tab-everywhere, but you commented specifically on a couple locations so I changed my code to be tab based. I could modify the whole file if you prefer. Thanks!

kpedro88 commented on Utils/src/HiddenSectorProducer.cc:
```diff
@@ -196,19 +236,162 @@ double HiddenSectorProducer::calculateMT2(const edm::Handle<edm::View<reco::GenM
+			int index = -1;
+			if ( it != pdgids.end()) {
+				index = it - pdgids.begin();
+				toMatch[index].genParts.push_back(genSubjetConstituents[i][j]);
+				toMatch[index].genIndexTop.push_back(i);
+				toMatch[index].genIndex.push_back(j);
```
the point is that these lines can be outside/after the if/else, because now they are exactly the same in both cases

easmith5 replied:
fantastic point

kpedro88 commented on Utils/src/HiddenSectorProducer.cc:
```diff
@@ -196,19 +236,162 @@ double HiddenSectorProducer::calculateMT2(const edm::Handle<edm::View<reco::GenM
+		int index = -1;
+		if ( it != pdgids.end()) {
+			index = it - pdgids.begin();
+			toMatch[index].recoParts.push_back(jetCands[r]);
+			toMatch[index].recoIndex.push_back(r);
```
and here

kpedro88 commented:
The spacing looks consistent enough now

Commit:
fix more things

kpedro88 commented on Utils/src/HiddenSectorProducer.cc:
```diff
@@ -367,13 +618,225 @@ void HiddenSectorProducer::produce(edm::StreamID, edm::Event& iEvent, const edm:
+		  std::vector<bool> recoMatched(jets_cands[recoJetIndex].size(), -1);
+		  std::vector<int> matchStageReco(jets_cands[recoJetIndex].size(), -1);
+
+		  std::vector<std::vector<int>> matchedRecoIndex = matchParticles(GenJets_darkHadronJets_constituents->at(genJetIndex), GenJets_darkHadronJets_constituentsPdgid->at(genJetIndex), jets_cands[recoJetIndex], cands_pdgids[recoJetIndex], recoMatched, matchStageReco);
```
`const std::vector<std::vector<int>>&`

kpedro88 commented on Utils/src/HiddenSectorProducer.cc:
```diff
@@ -196,19 +236,159 @@ double HiddenSectorProducer::calculateMT2(const edm::Handle<edm::View<reco::GenM
+			const auto& daughterPtrs = jet.daughterPtrVector();
+			if(cands_out[j].size()==daughterPtrs.size()) continue;
+			bool keep = false;
+			for(const auto& candJetDau : jet.daughterPtrVector()){
```
use the existing `daughterPtrs` variable here instead of calling `daughterPtrVector()` again (costly)

kpedro88 commented on Utils/src/HiddenSectorProducer.cc:
```diff
@@ -196,19 +236,159 @@ double HiddenSectorProducer::calculateMT2(const edm::Handle<edm::View<reco::GenM
+			if (keep) break;
+		}
+	}
+	return;
```
unneeded

kpedro88 commented:
I ran a profile on our benchmark signal sample.

Before:
* `HiddenSectorProducer` is 0.89 / 269.88 CPU ticks
    ```
    [671]       0.3       0.89       0.02 / 0.87       HiddenSectorProducer::produce(edm::StreamID, edm::Event&, edm::EventSetup const&) const
                0.2  .........       0.63 / 0.63         HiddenSectorProducer::checkLast(reco::GenJet const&, std::unordered_set<reco::Candidate const*, std::hash
                0.0  .........       0.13 / 0.13         HiddenSectorProducer::firstDark(reco::Candidate const*, std::unordered_set<reco::Candidate const*, std::h
                0.0  .........       0.04 / 0.22         HiddenSectorProducer::isParticle(std::unordered_set<unsigned int, std::hash<unsigned int>, std::equal_to<
   ```

* Size/Entry: 9.11 KiB

After:
* `HiddenSectorProducer` is 12.41 / 279.34 CPU ticks (6th largest contributor)
    ```
    [68]        3.6      12.41       0.05 / 12.36      HiddenSectorProducer::produce(edm::StreamID, edm::Event&, edm::EventSetup const&) const
                2.5  .........       8.57 / 8.57         NjettinessHelper::getTau(unsigned int, std::vector<ROOT::Math::LorentzVector<ROOT::Math::PtEtaPhiE4D<doub
                0.4  .........       1.32 / 1.32         HiddenSectorProducer::matchParticles(std::vector<std::vector<ROOT::Math::LorentzVector<ROOT::Math::PtEtaP
                0.2  .........       0.77 / 0.77         HiddenSectorProducer::matchJetsCands(edm::Handle<edm::View<pat::Jet> >&, edm::Handle<edm::View<reco::Cand
                0.2  .........       0.52 / 0.52         HiddenSectorProducer::checkLast(reco::GenJet const&, std::unordered_set<reco::Candidate const*, std::hash
                0.1  .........       0.38 / 0.38         HiddenSectorProducer::getAncestor(std::unordered_set<unsigned int, std::hash<unsigned int>, std::equal_to
                0.0  .........       0.16 / 0.16         HiddenSectorProducer::firstDark(reco::Candidate const*, std::unordered_set<reco::Candidate const*, std::h
    ```
* Size/Entry: 15.06 KiB

The main contributor is Nsubjettiness, which we can't really speed up. The biggest part of `matchParticles()` seems to be just the `matchAB()` calls, which are also needed. I found a few minor speedups here and there by looking at these parts of the code a little more closely.

The new branches ordered by size:
```
GenJetsAK8_darkHadronJets_constituents
JetsAK8_darkHadronJets_constituentsNextDH
JetsAK8_darkHadronJets_constituentsGenMatchOnly
JetsAK8_darkHadronJets_constituents
GenJetsAK8_darkHadronJets
GenJetsAK8_darkHadronJetsTau1
GenJetsAK8_darkHadronJetsTau2
GenJetsAK8_darkHadronJetsTau3
JetsAK8_darkHadronJets
GenJetsAK8_darkHadronJetsMultiplicity
```
Is there anything that could be dropped or condensed?

easmith5 commented:
I will add these changes. Indeed as you say, I'm not sure we can do much NsubjettinessHelper. For the size of the branches, there probably is some condensing that could be done. 

```
GenJetsAK8_darkHadronJets_constituents - gen particles clustered by dark hadron ancestors
JetsAK8_darkHadronJets_constituentsNextDH - matched particles + assigned particles to the closest dark hadron 
JetsAK8_darkHadronJets_constituentsGenMatchOnly - matched particles only
JetsAK8_darkHadronJets_constituents  - matched particles + assigned particles
GenJetsAK8_darkHadronJets - dark hadron jets computed from all daughters 
GenJetsAK8_darkHadronJetsTau1 
GenJetsAK8_darkHadronJetsTau2
GenJetsAK8_darkHadronJetsTau3
JetsAK8_darkHadronJets - dark hadron jets computer from all candidates matched to gen dark hadron daughters
GenJetsAK8_darkHadronJetsMultiplicity - number of constituents which are daughters of each dark hadron
``` 

It is very much the case that `JetsAK8_darkHadronJets_constituentsNextDH`, `JetsAK8_darkHadronJets_constituentsGenMatchOnly`, `JetsAK8_darkHadronJets_constituents` could be condensed, as the GenMatchOnly is essentially a subset of the other two. I could instead of storing the full lists again, store only the additional particles, and then combine them in coffea afterwards. I think it is useful to keep all of the different configurations available to be accessed separately as it may be useful for debugging, or validation of the process. The `JetsAK8_darkHadronJets`/`GenJetsAK8_darkHadronJets` is required by coffea to be able to access the `*_darkHadronJets_constituents*` branches properly with the coffea nesting. 

Ultimately non of the Tau branches are required for the actual reweighting process but I think they will be necessary for the documentation of the procedure. 

`GenJetsAK8_darkHadronJetsMultiplicity` could probably be dropped, one can just look at the length along the innermost axis of the constituent branches to get the number of constituents in the darkHadronJets. 

I suppose there is a workflow in which we store only the indices of the constituent particles and this would reduce size greatly (similar to other producers), but introduce some complexity into the following steps. Let me know how you think I should proceed - a 40% increase is fairly significant.

kpedro88 commented:
Here are some possibilities to reduce the disk usage:
1. Just store indices into the existing JetsConstituents and GenParticles branches. This would mean HiddenSectorProducer would consume some output from JetsConstituents and GenParticlesProducer. Probably, the best approach would be for those producers to output transient vectors (i.e. not stored in the final output file) that map the index in the miniAOD collection to the index in the TreeMaker branch. (Then HiddenSectorProducer can just track the indices during its loops.)
2. If the above seems like it may take too long to implement right away, some duplication could still be reduced by storing just one set of `JetsAK8_darkHadronJets_constituents` along with some boolean masks or integers to keep track of the different properties (match stage, which dark hadron jet they're assigned to). One could even combine this option with the first option to further reduce duplication.

These would both require some changes in how everything is reassembled in coffea. But that might actually improve the analysis processing (just loading one set of constituents in memory, instead of several copies of the same constituents).

One more possibility for Nsubjettiness: it is now actually exposed in the Python version of fastjet, as of the latest release https://github.com/scikit-hep/fastjet/releases/tag/v3.4.3.1. So we could eschew that entirely in TreeMaker and just run it as a spot check on the final output...

easmith5 commented:
For 1. I think this is certainly possible, but interacting with the output of other producers and validating the procedure would definitely slow things down, as I've already spent some time validating the existing setup. 

2 is definitely doable fairly quickly. I've basically implemented it already and I just want to do some checks on the output. I will also check that Nsubjettiness works and looks as expected in the python version of fastjet.  

2 is fairly easy to adjust for on the coffea side, it just requires implementing a couple of masks.  1. would potentially be a bit more difficult though ultimately it is also probably just a matter of reindexing/masking, though perhaps a more delicate one.

Commit:
Reduce output of lund code, add additional branches for masking purposes, fix bugs, and some small inefficiencies

easmith5 commented:
Okay. I have reduced the output size significantly. I didn't run a full profile, but just comparing output file sizes, this has helped dramatically. Removing nsubjettinesshelper should also decrease the time as well. Let me know if you see any additional issues!

kpedro88 commented on TreeMaker/interface/TreeMaker.h:
```diff
@@ -222,7 +223,7 @@ typedef TreeObject<vector<vector<math::PtEtaPhiELorentzVector>>,vector<vector<ma
 typedef TreeObject<vector<vector<math::XYZVector>>,vector<vector<math::XYZVectorF>>> TreeObjectVVXYZVToF;
 typedef TreeObject<vector<vector<math::XYZPoint>>,vector<vector<math::XYZPointF>>> TreeObjectVVXYZPToF;
-//typedefs for doubles
+//typedefs for doublesx
```
fix this typo again

kpedro88 commented on Utils/src/HiddenSectorProducer.cc:
```diff
@@ -80,13 +79,10 @@ class HiddenSectorProducer : public edm::global::EDProducer<edm::StreamCache<Nje
     double coneSize_;
     edm::ParameterSet njhConfig_;
     PidSet DarkSMediatorIDs_, DarkTMediatorIDs_, DarkQuarkIDs_, DarkHadronIDs_, DarkGluonIDs_, DarkStableIDs_, DarkFirstIDs_, SMQuarkIDs_;
-    std::unique_ptr<NjettinessHelper> beginStream(edm::StreamID) const {
```
to be sure, the usage of the Python FastJet Nsubjettiness has been validated now?

easmith5 replied:
So it hasn't yet been fully validated due to the coffea issues. I will do an initial check as best I can soon.

kpedro88 commented on Utils/src/HiddenSectorProducer.cc:
```diff
@@ -259,11 +254,8 @@ void HiddenSectorProducer::matchJetsCands(edm::Handle<edm::View<pat::Jet>>& h_je
 					break;
 				}
 			}
-			// if we are here either we broke out of candJetDau loop because we matched that jet dau, and we should skip the rest of the jets by breaking out of jet loop and going to next h_cand, or we went through all jet daughters and keep is false and we should go to next jet and see if we match there. Basically if we break out of daughters, also break out of jets loop
```
why was this removed?

easmith5 replied:
Ah I thought this was request - I guess you just meant the return statement?

kpedro88 replied:
Yes, just the very last return statement was unnecessary. I think skipping the loop breaking is incorrect.

kpedro88 commented on Utils/src/HiddenSectorProducer.cc:
```diff
@@ -326,14 +317,16 @@ std::vector<std::vector<int> > HiddenSectorProducer::matchParticles(std::vector<
 
-	for( auto& m: toMatch ) {
+	for( unsigned j = 0; j < toMatch.size(); j++ ) {
+		matchInfo m = toMatch[j];
```
`matchInfo& m` to avoid copying

kpedro88 commented on Utils/src/HiddenSectorProducer.cc:
```diff
@@ -792,17 +765,24 @@ void HiddenSectorProducer::produce(edm::StreamID iID, edm::Event& iEvent, const
+					  Jets_darkHadronJets_constituentsAssignedFirst->at(recoJetIndex)[idx[0]].push_back(1);
+					  Jets_darkHadronJets_constituentsAssignedSecond->at(recoJetIndex)[idx[0]].push_back(0);
+					  Jets_darkHadronJets_constituentsPdgid->at(recoJetIndex)[idx[0]].push_back(cands_pdgids[recoJetIndex][i]);
+					  Jets_darkHadronJets_constituentsMatchStage->at(recoJetIndex)[idx[0]].push_back(3);
```
is there a legend describing the match stage values somewhere? (preferably, these should be written as named constants rather than magic numbers)

easmith5 replied:
I haven't added a legend somewhere. When you say named constants you just mean something like: 
```
const int egMatchStage = 0
const int hadronMatchStage = 1
```
etc?

kpedro88 replied:
yes (or an enum, either way)

easmith5 replied:
Is putting this with the typedefs at the top of the file appropriate?

kpedro88 replied:
Best to make them members of the class to avoid the (slight) possibility of global contamination

Commit:
Fix a couple things, add const int for match stages

kpedro88 commented:
New profile:
* `HiddenSectorProducer` is 3.69 / 269.22 CPU ticks
    ```
    [178]       1.1       3.69       0.08 / 3.60       HiddenSectorProducer::produce(edm::StreamID, edm::Event&, edm::EventSetup const&) const
                0.4  .........       1.33 / 1.33         HiddenSectorProducer::matchParticles(std::vector<std::vector<ROOT::Math::LorentzVector<ROOT::Math::PtEtaP
                0.2  .........       0.68 / 0.68         HiddenSectorProducer::matchJetsCands(edm::Handle<edm::View<pat::Jet> >&, edm::Handle<edm::View<reco::Cand
                0.2  .........       0.52 / 0.52         HiddenSectorProducer::checkLast(reco::GenJet const&, std::unordered_set<reco::Candidate const*, std::hash
                0.1  .........       0.32 / 0.32         HiddenSectorProducer::getAncestor(std::unordered_set<unsigned int, std::hash<unsigned int>, std::equal_to
                0.0  .........       0.12 / 0.12         HiddenSectorProducer::firstDark(reco::Candidate const*, std::unordered_set<reco::Candidate const*, std::h
    ```
* Size/Entry: 13.71 KiB

So there is some improvement in the size, but still a lot of growth vs. the baseline. It should be acceptable for now, but for Run 3 usage, the more parsimonious approach of just storing indices into a central PFCandidates collection should be implemented.

kpedro88 commented on Utils/src/HiddenSectorProducer.cc:
```diff
@@ -273,11 +477,48 @@ void HiddenSectorProducer::produce(edm::StreamID, edm::Event& iEvent, const edm:
+
+  auto GenJets_darkHadronJets_tau1 = std::make_unique<std::vector<std::vector<double>>>();
+  auto GenJets_darkHadronJets_tau2 = std::make_unique<std::vector<std::vector<double>>>();
+  auto GenJets_darkHadronJets_tau3 = std::make_unique<std::vector<std::vector<double>>>();
```
these lines can be removed since Nsubjettiness is no longer used

kpedro88 commented:
For the Run 3 rewrite, one possible plan (to reuse existing functionality):
1. output a collection of dark hadron jets (actual jet objects) and value maps of additional per-constituent properties (match stage etc.)
2. include that collection in the list of inputs to `JetsConstituents` (which already makes the lists of constituents by index for the regular jet collections) and allow access to the value maps
3. output a per-jet list of dark hadron jet indices (rather than the current nested vector of dark hadron jet 4-vectors)

so the final stored output would look like:
DarkHadronJets
DarkHadronJets_[property]
DarkHadronJets_constituentsIndex
JetsAK8_darkHadronJetsIndex
JetsConstituents_[property]
+ summary diagnostic info (NConstituentsUnmatched etc.)

Then the same thing should be done for GenParticles, but that will take a bit more work, since the GenParticle selection is more complicated... some `JetsConstituents`-like algorithm will have to be added to `GenParticlesProducer`. (Alternatively, `GenParticlesProducer` could be modified just to produce a mask for the entire collection, which could then be consumed by `GenJetsConstituents` as (mostly) a clone of `JetsConstituents`.)

Commit:
Fix bug, remove last bit of nsubjettiness code

easmith5 commented:
This sounds good to me. Checking the size from my end, I get the same value of ~13.7 kB / entry for the current implementation. I would like to note that the original DarkJets branch started with ~11.6 kB / entry. I agree that in the long term (for Run 3 iteration) it would be best to use an index based system, though my understanding was that much of the code would be restructured for Run 3 anyways. 

I am currently working on the check of fastjet nsubjettiness - this requires awkward >=2 which requires coffea >=2023. I am in the process of updating the newer coffea treemaker schema to be able to run fastjet nsubjettiness on the darkHadronJet constituents.  

Beyond this, everything looks good to me. I can run the full diagnostic and Lund Reweighting python pipeline on these outputs and get values for the per-jet weights.
